blockscout-ens/bens-logic/src/subgraphs_reader/reader.rs

```diff
@@ -0,0 +1,237 @@
+        &self,
+        _network_id: i64,
+        _name: &str,
+    ) -> Result<Vec<()>, SubgraphReadError> {
+        todo!()
+    }
+
+    pub async fn search_resolved_domain_reverse(
+        &self,
+        network_id: i64,
+        address: ethers::types::Address,
+    ) -> Result<Vec<Domain>, SubgraphReadError> {
+        let schema = self
+            .schema_names
+            .get(&network_id)
+            .ok_or_else(|| SubgraphReadError::NetworkNotFound(network_id))?;
+        let address = hex(address);
+        find_resolved_addresses(self.pool.as_ref(), schema, &address).await
+    }
+
+    pub async fn search_owned_domain_reverse(
+        &self,
+        network_id: i64,
+        address: ethers::types::Address,
+    ) -> Result<Vec<Domain>, SubgraphReadError> {
+        let schema = self
+            .schema_names
+            .get(&network_id)
+            .ok_or_else(|| SubgraphReadError::NetworkNotFound(network_id))?;
+        let address = hex(address);
+        find_owned_addresses(self.pool.as_ref(), schema, &address).await
+    }
+}
+
+async fn find_domain(
+    pool: &PgPool,
+    schema: &str,
+    id: &str,
+) -> Result<Option<Domain>, SubgraphReadError> {
+    let maybe_domain = sqlx::query_as(&format!(
+        r#"
+        SELECT
+        DISTINCT ON (id) *
+        FROM {schema}.domain
+        WHERE
+            id = $1 
+            AND name IS NOT NULL
+            AND to_timestamp(expiry_date) > now()
+        ORDER BY
+            id,
+            block_range DESC
+        "#,
+    ))
+    .bind(id)
+    .fetch_optional(pool)
+    .await?;
+    Ok(maybe_domain)
+}
+
+async fn find_resolved_addresses(
+    pool: &PgPool,
+    schema: &str,
+    address: &str,
+) -> Result<Vec<Domain>, SubgraphReadError> {
+    let resolved_domains: Vec<Domain> = sqlx::query_as(&format!(
+        r#"
+        SELECT * FROM (
+            SELECT 
+            DISTINCT ON (id) *
+            FROM {schema}.domain
+            WHERE 
+                resolved_address = $1
+                AND name IS NOT NULL 
+                AND to_timestamp(expiry_date) > now()
+            ORDER BY
+                id,
+                block_range DESC
+        ) sub
+        ORDER BY created_at ASC
+        "#,
+    ))
+    .bind(address)
+    .fetch_all(pool)
+    .await?;
+
+    Ok(resolved_domains)
+}
+
+async fn find_owned_addresses(
+    pool: &PgPool,
+    schema: &str,
+    address: &str,
+) -> Result<Vec<Domain>, SubgraphReadError> {
+    let owned_domains: Vec<Domain> = sqlx::query_as(&format!(
+        r#"
+        SELECT * FROM (
+            SELECT 
+            DISTINCT ON (id) *
+            FROM {schema}.domain
+            WHERE 
+                (
+                    owner = $1
+                    OR wrapped_owner = $1
+                )
+                AND name IS NOT NULL
+                AND to_timestamp(expiry_date) > now()
+            ORDER BY
+                id,
+                block_range DESC
```

Comment:
Q: Just interested, what `block_range` is used for in "domain" table?

Reply:
graph-node stores history of every entity using block_range. range `[1234,)` (right bound is null) means that this is current state of enitity